examples/kokkos/view/example_restrict.cpp

```diff
@@ -0,0 +1,243 @@
+#include "Kokkos_Core.hpp"
+#include "Kokkos_Profiling_ProfileSection.hpp"
+#include "Kokkos_Profiling_ScopedRegion.hpp"
+
+namespace reprospect::examples::kokkos::view
+{
+#define OPERATION(dst, src_a, src_b, index)    \
+    dst[index]  = src_a[index] + src_b[index]; \
+    dst[index] += src_a[index] + src_b[index];
+
+template <typename T>
+__global__ void global_kernel(
+    T* __restrict__ const dst,
+    const T* __restrict__ const src_a, const T* __restrict__ const src_b,
+    const unsigned int size
+) {
+    const auto index = blockIdx.x * blockDim.x + threadIdx.x;
+    if (index < size) {
+        OPERATION(dst, src_a, src_b, index)
+    }
+}
+
+template <typename T>
+struct FunctorRestrictMember
+{
+    const T* __restrict__ src_a;
+    const T* __restrict__ src_b;
+          T* __restrict__ dst;
+
+    template <std::integral U>
+    KOKKOS_FUNCTION void operator()(const U index) const {
+        OPERATION(dst, src_a, src_b, index)
+    }
+};
+
+template <typename T>
+struct FunctorRestrictRecast
+{
+    const T* src_a;
+    const T* src_b;
+          T* dst;
+
+    template <std::integral U>
+    KOKKOS_FUNCTION void operator()(const U index) const {
+        [index](T* __restrict__ const d, const T* __restrict__ const a, const T* __restrict__ const b) {
+            OPERATION(d, a, b, index)
+        }(dst, src_a, src_b);
+    }
+};
+
+template <typename T>
+struct FunctorRestrictView
+{
+    using view_t = Kokkos::View<T*, Kokkos::CudaSpace, Kokkos::MemoryTraits<Kokkos::Restrict>>;
+
+    typename view_t::const_type src_a, src_b;
+    view_t dst;
+
+    template <std::integral U>
+    KOKKOS_FUNCTION void operator()(const U index) const {
+        OPERATION(dst, src_a, src_b, index)
+    }
+};
+
+//! Inspired by https://github.com/kokkos/kokkos/blob/37f70304dc3676691af88d3ac3ba50cddbfa337f/core/src/Cuda/Kokkos_Cuda_View.hpp#L17-L26.
+template <typename T> requires (sizeof(T) == 4)
+struct LDGAccessor
+{
+    const T* m_ptr;
+
+    template <typename U>
+    KOKKOS_FUNCTION T operator[](const U index) const {
+        return __ldg(&m_ptr[index]);
+    }
+};
+
```

Comment:
We could also move this "LDG" case to the end. We could have
- global kernel
- restrict cases that work (function params, pointers in local scope)
- restrict cases that don't work (member, trait, accessor)
- ldg